### ISSUE TITLE
PowerShell SDK: Opening an out-of-process runspace fails

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#8652]
Note: This is a regression from Windows PowerShell.

The error occurs when `.Open()` is called.

Note that equivalent code written in and run from PowerShell does _not_ exhibit this problem (#8643  shows an example).

# Steps to reproduce

Create a .NET Core console application with the following code and run it with `dotnet run`:

```csharp
using System;
using System.Management.Automation;
using System.Management.Automation.Runspaces;

namespace Sample
{
    class Program
    {
        static void Main(string[] args)
        {
            using (var runspace = RunspaceFactory.CreateOutOfProcessRunspace(null))
            {
              runspace.Open();
            }
        }
    }
}
```

# Expected behavior

The application should run and exit quietly, with exit code `0`.

# Actual behavior

The exit code is `134`, and the following error message is printed:

```none

Unhandled Exception: System.Management.Automation.Remoting.PSRemotingTransportException: An error occurred while starting the background process. Error reported: No such file or directory. ---> System.ComponentModel.Win32Exception: No such file or directory
   at Interop.Sys.ForkAndExecProcess(String filename, String[] argv, String[] envp, String cwd, Boolean redirectStdin, Boolean redirectStdout, Boolean redirectStderr, Boolean setUser, UInt32 userId, UInt32 groupId, Int32& lpChildPid, Int32& stdinFd, Int32& stdoutFd, Int32& stderrFd, Boolean shouldThrow)
   at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()
   at System.Management.Automation.Runspaces.PowerShellProcessInstance.Start()
   at System.Management.Automation.Remoting.Client.OutOfProcessClientSessionTransportManager.CreateAsync()
   --- End of inner exception stack trace ---
   at System.Management.Automation.Runspaces.AsyncResult.EndInvoke()
   at System.Management.Automation.Runspaces.Internal.RunspacePoolInternal.EndOpen(IAsyncResult asyncResult)
   at System.Management.Automation.Runspaces.Internal.RemoteRunspacePoolInternal.Open()
   at System.Management.Automation.Runspaces.RunspacePool.Open()
   at System.Management.Automation.RemoteRunspace.Open()
   at Sample.Program.Main(String[] args) in /path/to/Program.cs:line 13
```

Additionally, the `dotnet` host may crash.

# Environment data

```none
Targets netcoreapp2.1 and netcoreapp2.2 with
Microsoft.PowerShell.SDK NuGet package v6.2.0-preview.3 on:
macOS 10.14.2
Ubuntu 18.04.1 LTS
Microsoft Windows 10 Pro (64-bit; Version 1803, OS Build: 17134.471)
```

**Comment by MaximoTrinidad** [comment]
Just wondering!!

Which version of the PowerShell SDK we are talking about??

:)

**Comment by mklement0** [comment]
Good point, @MaximoTrinidad; I've modified the "Environment data" section to state the NuGet package version.

**Comment by Sjark** [comment]
This also happens when targeting netcoreapp3.0 with nuget package:
System.Management.Automation version 6.2.3

This seems to happen cause it can't find the pwsh.exe file.
In my project, it looks for pwsh.exe in folder: 
"bin\Debug\netcoreapp3.0\runtimes\win-x64\lib\netcoreapp2.1", 

but the only files that are there is: 
"Microsoft.PowerShell.CoreCLR.Eventing.dll"
"System.Management.Automation.dll".

It would be nice if you had the option to specifiy the pwsh.exe path when creating a new PowerShellProcessInstance, that way you can use a version of pwsh already deployed to the machine you are running on.

**Comment by LunicLynx** [comment]
@Sjark 
Just found this hack to set another path
```csharp
var pspi = new PowerShellProcessInstance();
pspi.Process.StartInfo.FileName = @"C:\Windows\System32\WindowsPowerShell
pspi.Process.Start();
var runspace = RunspaceFactory.CreateOutOfProcessRunspace(new TypeTable(new string[0]), pspi);
```

**Comment by Sjark** [comment]
Just tested the workaround @LunicLynx provided using System.Management.Automation@7.0.0-preview.4.

This worked great when pointing to pwsh.exe 7.0.0-preview.4.

I also tried targeting Windows Powershell like in the comment above, this failed cause it tried to start powershell 2.0 for some reason.

This can be fixed by adding:
```csharp
pspi.Process.StartInfo.Arguments = $" -Version 5.1{pspi.Process.StartInfo.Arguments}";
```

This workaround works great, but it would be nice to be able to configure this when creating new PowerShellProcessInstance.

**Comment by leonardder** [comment]
This problem still applies to PowerShell 7.0.0 final.

**Comment by IISResetMe** [comment]
Just ran into this in 7.0.1 as well. 

Manually setting `Process.StartInfo.FileName` to "C:\Program Files\PowerShell\7\pwsh.exe" (still) works on Windows, but this will be a mess to monkey-patch in a serverless deployment 😒

**Comment by arunswarnam** [comment]
`pspi.Process.StartInfo.FileName = @"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe";` . Is there a way for this workaround to be able to run on all platforms supported by PSCore by detecting the path at execution rather than hard coding?

**Comment by IISResetMe** [comment]
The runtime detection _is_ the problem:

https://github.com/PowerShell/PowerShell/blob/5f28df120ab95ebb8871d19d3bdb8046c5b9817c/src/System.Management.Automation/engine/Utils.cs#L490-L492

Not sure what the most robust method for detecting `$PSHome` for a specific version at runtime would be

**Comment by Sjark** [comment]
If locating `$PSHome` for a given version at runtime is the problem, there should at least be an option to set it in the constructor of `PowerShellProcessInstance`

**Comment by ZazioMDT** [comment]
I just ran into this using 7.0.3.  I fixed it by passing in 5.1 in the version field. 

            using (PowerShellProcessInstance instance = new PowerShellProcessInstance
                (new Version(5, 1), null, null, true))
            {
                using (Runspace rs = RunspaceFactory.CreateOutOfProcessRunspace(
                    new TypeTable(Array.Empty<string>()), instance))
                {

I tried out the StartInfo.FileName but then Write-Information was an unrecognized command.  This lead me to look into whether I was passing a correct version to the PowerShellProcessInstance constructor.  I had the Version at 5.0, I changed it to 5.1 and it started working.  I removed the FileName and it continued to work.
